### PR TITLE
SDK-960 - avoid nodes sync indefinitely

### PR DESCRIPTION
**Pull request by davehorizen**
## Description
On the current implementation there is a problem in the way we calculate and send `SyncInfo` to the other nodes. The other node height retrieved from the  `SyncInfo` message is not correctly estimated and this lead in some cases to consider another node more updated as younger. In this situation the node consider itself more developed and stop processing other `SyncInfo` messages remaining stuck at certain height.
With the current PR we introduced a new method to correctly calculate the other node height (method `calculateOtherNodeApproxHeight`).
The Equal return statement in case the divergent suffix sequence size is greater than one was  removed because it is currently a fork situation, an existing comment on the code reported exactly that.

## Jira Ticket
details in task [SDK-960](https://horizenlabs.atlassian.net/browse/SDK-960)

## Checks
- [X] Project Builds
- [X] Project passes tests and checks
- [ ] Updated documentation accordingly

[SDK-960]: https://horizenlabs.atlassian.net/browse/SDK-960?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 1]
```diff
@@ -368,20 +388,14 @@ abstract class AbstractHistory[
...
-        else if (storage.height == otherBestBlockApproxHeight) {
-          if(otherBestBlockApproxHeight == otherBestKnownBlockHeight)
-            Equal // UPDATE: FORK in both cases
-          else
-            Fork
-        }
+        else if (storage.height == otherBestBlockApproxHeight)
+          Fork
         else
           Younger
```
Comment (i-skrypnyk) [inline sdk/src/main/scala/io/horizen/history/AbstractHistory.scala:400]: Do you think it is safe to set the status to Younger as default here?
This case means that that other node has fewer blocks, but there is also a Fork, because a number of blocks from the other node history was not found in local history.
So it is both Fork and Younger

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 2]
```diff
@@ -321,6 +320,28 @@ abstract class AbstractHistory[
     indexes :+ 1
   }
 
+  // calculate the other node approximate height knowing that the sequence was created using the method knownBlocksHeightToSync
+  protected def calculateOtherNodeApproxHeight(startingHeight: Int, sequenceSize: Int): Int = {
```
Comment (i-skrypnyk) [inline sdk/src/main/scala/io/horizen/history/AbstractHistory.scala:324]: Can you please add a test(unit will do) that will fail without the fix and pass with the fix

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 3]
```diff
@@ -341,6 +341,25 @@ abstract class AbstractHistory[
...
+    while(currentSequencePosition < (dSuffixSize-1)) {
+      if(currentSequencePosition < 10) {
+        heightAtIndex -= 1
+      } else {
+        step *= 2
+        heightAtIndex -= step
+      }
+      currentSequencePosition += 1
+    }
+    if(heightAtIndex < storage.height)
```
Comment (paologalligit) [inline sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: This can be simplified by only returning `heightAtIndex < storage.height`

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 4]
```diff
@@ -321,13 +320,46 @@ abstract class AbstractHistory[
...
+        currentSequencePosition -= 1
+      }
+      currentHeight
+    }
+  }
+
+  // check the presence of a block on the current history at the second divergent suffix sequence height
+  // it is used to check the presence of forks
+  private def checkForkAtPreviousIndexBlock(otherBestKnownBlockHeight: Int, dSuffixSize: Int): Boolean = {
+    var heightAtPreviousIndex = 0
```
Comment (paologalligit) [inline sdk/src/main/scala/io/horizen/history/AbstractHistory.scala:347]: Here we can directly assign
```
val heightAtPreviousIndex = if (dSuffixSize > 10) {
      otherBestKnownBlockHeight + math.pow(2, dSuffixSize-10).toInt
    } else {
      otherBestKnownBlockHeight + dSuffixSize
    } 
```

**Review by paologalligit — COMMENTED** [review]
I'd like to have a set of tests for the `checkForkAtPreviousIndexBlock` method corner cases

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala (hunk 1 above)
Comment (davehorizen) [reply sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: Good point, after further analysis I proceed to refactor the method logic.
The current logic in case the divergent suffix sequence size is greater than 1 is the following:
knowing our block height, the best known block height and the calculated best block related to the other node history, we do the following checks:
- if the other node best block height is greater than the current height:
	- and the height of the second index of the divergent suffix sequence is less then the current height we have a `Fork` (this because it means that the block id at a height less than our best block didn't found a match)
	- else the other node history is `Older`
the default condition for the other cases is `Fork`.

No need to manage the `Younger` case because in case the other node is really younger (with no forks) the divergent suffix sequence size will be 1 and we will be outside these cases

The `Unknown` condition is satisfied only if we are not able to calculate the other node best block height, in this case we don't have enough data to return a secure comparison result.

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala (hunk 2 above)
Comment (davehorizen) [reply sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: Unit test added (`SidechainHistoryTest.synchronizationTestWithHighBlockNumber`) to simulate the task example. The number of blocks involved is less than the one of the task to keep the test runtime low but enough to manage divergent suffix sequence that involved height difference with exponential step.

**Comment by davehorizen** [comment]
> I'd like to have a set of tests for the `checkForkAtPreviousIndexBlock` method corner cases

The unit test `SidechainHistoryTest.checkForkAtPreviousIndexBlock` is already covering this case [here](https://github.com/HorizenOfficial/Sidechains-SDK/blob/db/SDK-960/sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala#L634) and [here](https://github.com/HorizenOfficial/Sidechains-SDK/blob/db/SDK-960/sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala#L666) where we are comparing our history against a longer one with a fork at a certain height (a `Fork` case instead of an `Older` one)

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 5]
```diff
@@ -321,13 +320,46 @@ abstract class AbstractHistory[
...
+    if (dSuffixSize > 10) {
+      heightAtPreviousIndex = otherBestKnownBlockHeight + math.pow(2, dSuffixSize-10).toInt
+    } else {
+      heightAtPreviousIndex = otherBestKnownBlockHeight + dSuffixSize
+    }
+    if (heightAtPreviousIndex < storage.height)
```
[the remark this answers is not in the document]
Comment (davehorizen) [reply sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: agree, more clear! Done

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala (hunk 4 above)
Comment (davehorizen) [reply sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: done

**Comment by davehorizen** [comment]
Unit test pushed to cover a corner case of the` AbstractHistory.comparison` method that will result in an Older status even if there is a Fork between the two histories.
Details in the[ unit test comment](https://github.com/HorizenOfficial/Sidechains-SDK/blob/db/SDK-960/sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala#L615).

**Review by paologalligit — APPROVED** [review]

File: sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala [hunk 6]
```diff
@@ -534,29 +534,173 @@ class SidechainHistoryTest extends JUnitSuite
...
+    // -----------------------------------------------------------------------------------------------------------------
+    // Test 2:
+    // update history2 with 10 more blocks (introduce a fork).
+    // Here we can test a corner case in which the comparison between two histories will result in a Older state even if there is Fork
+    // The node1 (at height 300) will send the node2 a list with the modifiers ID at the following height:
+    // 300 - 299 - 298 - 297 - 296 - 295 - 294 - 293 - 292 - 291 - 289 - 285 - 277 - 261 - 229 - 165 - 37 - 1
+    // The divergent suffix list will be the modifiers at the following height (size 17):
+    // 37 - 165 - 229 - 261 - 277 - 285 - 289 - 291 - 292 - 293 - 294 - 295 - 296 - 297 - 298 - 299 - 300
+    // The internal logic will calculate the second divergent suffix height.
+    // The result (165) is less than the current height (110) and the final result will be an Older status
```
Comment (MarcoOl94) [inline sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala:623]:
```suggestion
    // The result (165) is greater than the current height (160) and the final result will be an Older status
```

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala [hunk 7]
```diff
@@ -321,13 +320,44 @@ abstract class AbstractHistory[
...
+          step *= 2
+          currentHeight += step
+        }
+        currentSequencePosition -= 1
+      }
+      currentHeight
+    }
+  }
+
+  // check the presence of a block on the current history at the second divergent suffix sequence height
```
Comment (paolocappelletti) [inline sdk/src/main/scala/io/horizen/history/AbstractHistory.scala:344]: let's change the comment and the name of the method to be more clear
checkForkAtSecondToLastDivergentSuffixIndex
or something similar :)

**Review by paolocappelletti — APPROVED** [review]

File: sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala (hunk 6 above)
Comment (davehorizen) [reply sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala]: fixed

File: sdk/src/main/scala/io/horizen/history/AbstractHistory.scala (hunk 7 above)
Comment (davehorizen) [reply sdk/src/main/scala/io/horizen/history/AbstractHistory.scala]: method name update, the one proposed seems fine and it was used!

File: sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala [hunk 8]
```diff
@@ -618,9 +618,9 @@ class SidechainHistoryTest extends JUnitSuite
     // The node1 (at height 300) will send the node2 a list with the modifiers ID at the following height:
     // 300 - 299 - 298 - 297 - 296 - 295 - 294 - 293 - 292 - 291 - 289 - 285 - 277 - 261 - 229 - 165 - 37 - 1
     // The divergent suffix list will be the modifiers at the following height (size 17):
-    // 37 - 165 - 229 - 261 - 277 - 285 - 289 - 291 - 292 - 293 - 294 - 295 - 296 - 297 - 298 - 299 - 300
+    // 1 - 37 - 165 - 229 - 261 - 277 - 285 - 289 - 291 - 292 - 293 - 294 - 295 - 296 - 297 - 298 - 299 - 300
```
Comment (MarcoOl94) [inline sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala:621]: Are you sure of this 1?

**Review by MarcoOl94 — APPROVED** [review]

File: sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala (hunk 8 above)
Comment (davehorizen) [reply sdk/src/test/scala/io/horizen/utxo/integration/SidechainHistoryTest.scala]: removed
